src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -0,0 +1,144 @@
+    (let [v4-accept-val "binary"
+          vals (str/split accept #"\s*,\s*")                ; BAD! see RFC-2616
+          munged-accept (->> vals
+                          (replace {"s" v4-accept-val "pson" v4-accept-val})
+                          (distinct)
+                          (str/join ", "))]
+      (assoc-in request [:headers "accept"] munged-accept))
+    request))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Handler wrapper
+
+(defn request-compatibility-wrapper
+  "Given a Puppet 3.X CA request, convert it to a Puppet 4.X CA request."
+  [handler mount-point api-version request]
+  (let [{{environment :environment} :params
+         path-info                  :path-info
+         query-string               :query-string} request
+        query-params (if query-string (ring-codec/form-decode query-string) {})]
+    (let [compat-request
+          (-> request
+              (map-accept-raw-to-binary)
+              (assoc :path-info    (str "/" api-version path-info)
+                     :context      mount-point
+                     :uri          (str mount-point "/" api-version path-info)
+                     :query-string (ring-codec/form-encode
+                                     (merge
+                                       query-params
+                                       (when environment {:environment environment}))))
+              (dissoc :params :route-params))]
+      (handler compat-request))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Routing
+
+(defn v2_0-routes
+  "Creates the compojure routes to handle the 3.X master's '/v2.0' routes."
+  [request-handler]
+  (compojure/routes
+    (compojure/GET "/environments" request
+                   (request-handler request))))
+
+(defn legacy-master-routes
+  [master-request-handler ca-request-handler]
+  (compojure/routes
+    (compojure/GET "/node/*" request (master-request-handler request))
+    (compojure/GET "/facts/*" request (master-request-handler request))
+    (compojure/GET "/file_content/*" request (master-request-handler request))
+    (compojure/GET "/file_metadatas/*" request (master-request-handler request))
+    (compojure/GET "/file_metadata/*" request (master-request-handler request))
+    (compojure/GET "/file_bucket_file/*" request
+      (master-request-handler (map-accept-s-pson-to-binary request)))
+    ;; TODO: file_bucket_file request PUTs from Puppet agents currently use a
+    ;; Content-Type of 'text/plain', which, per HTTP specification, would imply
+    ;; a default character encoding of ISO-8859-1 or US-ASCII be used to decode
+    ;; the data.  This would be incorrect to do in this case, however, because
+    ;; the actual payload is "binary".  Coercing this to
+    ;; "application/octet-stream" for now as this is synonymous with "binary".
+    ;; This should be removed when/if Puppet agents start using an appropriate
```

Comment:
I think we could just take the "TODO" and the "This should be removed when/if..." part off now because we're adding this back in specifically to support Puppet 3.X agent behavior which won't change in newer Puppet 3.X releases, for compatibility reasons.